src/generate.js

```diff
@@ -2,39 +2,89 @@ import Prefixer from 'inline-style-prefixer';
 
 import {
     objectToPairs, kebabifyStyleName, recursiveMerge, stringifyValue,
-    importantify
+    importantify,
 } from './util';
 
+// In a bundle of styles, find all of the different class names that a single
+// named descendant selector can have.
+const findNamesForDescendants = (styles, names) => {
+    Object.keys(styles).forEach(key => {
+        if (key[0] === ':' || key[0] === '@') {
+            // Recurse for pseudo or @media styles
+            findNamesForDescendants(styles[key], names);
+        } else if (key[0] === '>' && key[1] === '>') {
+            // Recurse for descendant styles
+            findNamesForDescendants(styles[key], names);
+
+            // Pluck out all of the names in the _names object.
+            Object.keys(styles[key]._names).forEach(name => {
+                names[key] = names[key] || [];
+                names[key].push(name);
+            });
+        }
+    });
+};
+
 export const generateCSS = (selector, styleTypes, stringHandlers,
-        useImportant) => {
+                            useImportant) => {
     const merged = styleTypes.reduce(recursiveMerge);
 
+    const classNamesForDescendant = {};
+    findNamesForDescendants(merged, classNamesForDescendant);
+
+    return generateCSSInner(
+        selector, merged, stringHandlers, useImportant,
+        classNamesForDescendant);
+}
+
+const generateCSSInner = (selector, style, stringHandlers,
+                          useImportant, classNamesForDescendant) => {
     const declarations = {};
     const mediaQueries = {};
+    const descendants = {};
     const pseudoStyles = {};
 
-    Object.keys(merged).forEach(key => {
+    Object.keys(style).forEach(key => {
         if (key[0] === ':') {
-            pseudoStyles[key] = merged[key];
+            pseudoStyles[key] = style[key];
         } else if (key[0] === '@') {
-            mediaQueries[key] = merged[key];
+            mediaQueries[key] = style[key];
+        } else if (key[0] === '>' && key[1] === '>') {
+            // So we don't generate weird "_names: [Object object]" styles,
+            // make a copy of the styles and get rid of the _names value.
+            const stylesWithoutNames = {
```

Comment:
You could avoid the use of `delete` (which has negative perf implications) by doing this:

```javascript
const {_names, ...stylesWithoutNames} = style[key]
```

[For example](http://babeljs.io/repl/#?evaluate=true&presets=es2015%2Creact%2Cstage-2&experimental=false&loose=false&spec=false&code=const%20style%20%3D%20%7B%0A%20%20bar%3A%20%7B%0A%20%20%20%20_names%3A%20'blah'%2C%0A%20%20%20%20otherProperty%3A%201%2C%0A%20%20%20%20otherProp%3A%202%2C%0A%20%20%7D%0A%7D%0A%0Aconst%20key%20%3D%20'bar'%0A%0A%2F%2F%20the%20relevant%20code%3A%0Aconst%20%7B_names%2C%20...stylesWithoutNames%7D%20%3D%20style%5Bkey%5D%0Aconsole.log(stylesWithoutNames))

:-)